app/models/papi_client.rb

```diff
@@ -324,6 +351,54 @@ def get_command_line_options(study_file, action)
     opts
   end
 
+  # get command line arguments for differential expression runs
+  #
+  # * *params*
+  #   - +options+ (Hash) => Hash of DE-specific parameters, whose names must include all values from DE_PARAMETER_NAMES
+  #
+  # * *returns*
+  #   - (Array) => Array of DE-specific command line arguments to pass to Docker in "exec" form
+  #
+  # * *raises*
+  #   - (ArgumentError) => if any required parameters from DE_PARAMETER_NAMES are missing or malformed
+  def get_de_parameters(options)
+    # determine list of params to validate with by detecting matrix type, with a default of 'dense'
+    matrix_type = options.detect { |name, _| name.to_s == 'matrix_file_type' }&.last || 'dense'
+    control_params = matrix_type == 'sparse' ? DE_PARAMETER_NAMES + SPARSE_DE_PARAMS : DE_PARAMETER_NAMES
+
+    # validate input parameters
+    safe_params = options.select { |name, _| control_params.include?(name.to_s) }
+    missing_params = control_params - safe_params.keys.map(&:to_s)
+    if missing_params.any?
+      raise ArgumentError, "cannot run differential expression due to missing parameters: #{missing_params.join(', ')}"
+    end
+
+    # add all good parameters to params array so they can be deserialized as command line options for Docker
+    # will throw an exception if value passed is invalid for the given parameter type
+    de_params = []
+    safe_params.each do |param_name, value|
+      if %w[cluster_file annotation_file matrix_file_path].include?(param_name)
+        raise ArgumentError, "#{param_name} contains a non-gs URL value: #{value}" unless value.start_with?('gs://')
+      end
+
+      de_params += [to_cli_opt(param_name), value]
+    end
+    de_params << '--differential-expression' # final argument that has no value
+    de_params
+  end
+
+  # transform a parameter from Hash key or String into a representation that can be called in the Python CLI
+  # converts underscores (_) to dashes (-) and prepends double dashes (--)
+  #
+  # * *params*
+  #   - +param_name+ (Symbol, String) => Name of parameter
+  #
+  # * *returns*
+  #   - (String) => New parameter name, encoded as kwarg for Python command line
+  def to_cli_opt(param_name)
```

Comment:
For easier testing without having to instantiate the class
```suggestion
  def self.to_cli_opt(param_name)
```

Reply:
[8f7fef6](https://github.com/broadinstitute/single_cell_portal_core/pull/1483/commits/8f7fef66d1357c74e4dea90ce08a41f8fb81ae98)